// vcc-ui/lib/story-wrapper.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ThemeWrapper = ThemeWrapper;
exports.StoryWrapper = StoryWrapper;

var _react = _interopRequireDefault(require("react"));

var _addonKnobs = require("@storybook/addon-knobs");

var _reactFela = require("react-fela");

var _styleRenderer = require("./style-renderer");

var _volvo = _interopRequireDefault(require("./themes/volvo"));

var _polestar = _interopRequireDefault(require("./themes/polestar"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var renderer = (0, _styleRenderer.styleRenderer)();
var themes = {
  volvo: _volvo.default,
  polestar: _polestar.default
};
var currentTheme = "volvo";

function ThemeWrapper(_ref) {
  var children = _ref.children;
  var theme = (0, _addonKnobs.select)("theme", Object.keys(themes), currentTheme);
  currentTheme = theme;
  var rtl = (0, _addonKnobs.boolean)("rtl", false);
  var dir = rtl ? "rtl" : "ltr";
  document.body.setAttribute("dir", dir);
  return _react.default.createElement(_reactFela.ThemeProvider, {
    theme: _objectSpread({}, themes[theme], {
      logoImagesPath: "/static/vcc-ui/images/",
      fontsPath: "/static/vcc-ui/fonts/",
      direction: dir
    })
  }, children);
}

function StoryWrapper(_ref2) {
  var children = _ref2.children;
  return _react.default.createElement(_reactFela.RendererProvider, {
    renderer: renderer
  }, _react.default.createElement(ThemeWrapper, null, children));
}
