// vcc-ui/lib/components/logo/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Logo = Logo;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _reactFela = require("react-fela");

var _block = require("../block");

var _helpers = require("./helpers");

var _getThemeStyle = require("../../get-theme-style");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Logo(_ref) {
  var type = _ref.type,
      height = _ref.height,
      altText = _ref.altText;

  var _useFela = (0, _reactFela.useFela)(),
      theme = _useFela.theme;

  var _getImagesSrc = (0, _helpers.getImagesSrc)({
    type: type,
    theme: theme
  }),
      src = _getImagesSrc.src,
      srcSet = _getImagesSrc.srcSet;

  return _react.default.createElement(_block.Block, {
    as: "img",
    src: src,
    srcSet: srcSet,
    alt: altText,
    height: height,
    extend: {
      verticalAlign: "middle",
      maxHeight: "100%",
      // this allows passing a function of theme
      // as a generic extension via theme
      extend: (0, _getThemeStyle.getThemeStyle)("logo", theme)
    }
  });
}

Logo.propTypes = {
  /** Logo type */
  // type: PropTypes.oneOf([LOGO_TYPES.SQUARE, LOGO_TYPES.WORDMARK]),
  type: _propTypes.default.string,

  /** Apply custom height to logo image */
  height: _propTypes.default.number,

  /** Alternative text for the logo image */
  altText: _propTypes.default.string
};
