// vcc-ui/es/components/click/index.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

function _objectWithoutProperties(source, excluded) { if (source == null) return {}; var target = _objectWithoutPropertiesLoose(source, excluded); var key, i; if (Object.getOwnPropertySymbols) { var sourceSymbolKeys = Object.getOwnPropertySymbols(source); for (i = 0; i < sourceSymbolKeys.length; i++) { key = sourceSymbolKeys[i]; if (excluded.indexOf(key) >= 0) continue; if (!Object.prototype.propertyIsEnumerable.call(source, key)) continue; target[key] = source[key]; } } return target; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

import React from "react";
import PropTypes from "prop-types";
import { Block } from "../block";
var buttonResetStyle = {
  background: "none",
  margin: 0,
  padding: 0,
  textAlign: "left",
  cursor: "pointer",
  appearance: "none",
  borderWidth: 0,
  "::-moz-focus-inner": {
    borderWidth: 0,
    padding: 0
  }
};
var linkResetStyle = {
  textDecoration: "none"
};
export function Click(_ref) {
  var children = _ref.children,
      extend = _ref.extend,
      innerRef = _ref.innerRef,
      props = _objectWithoutProperties(_ref, ["children", "extend", "innerRef"]);

  var as = props.href ? "a" : "button";
  return React.createElement(Block, _extends({}, props, {
    innerRef: innerRef,
    as: as,
    type: as === "button" ? props.type || "button" : null,
    extend: [{
      boxSizing: "border-box",
      color: "#000",
      fontSize: "12px",
      fontFamily: "sans-serif"
    }, as === "button" ? buttonResetStyle : linkResetStyle, extend]
  }), children);
}
Click.propTypes = {
  /** Any valid React element, function, or a string specifying a name for an HTML element */
  as: PropTypes.oneOfType([PropTypes.string, PropTypes.element, PropTypes.func]),

  /** An object containing valid CSS style declarations */
  extend: PropTypes.oneOfType([PropTypes.object, PropTypes.func, PropTypes.array]),

  /** A JSX node */
  children: PropTypes.node,

  /** React DOM ref object or fn to be passed through to the DOM element **/
  innerRef: PropTypes.oneOfType([PropTypes.object, PropTypes.func])
};
